File: 2 Regression/E Decision Tree/my_decision_tree_regression.py
# -*- coding: utf-8 -*-
"""
Created on Wed Jun 20 23:40:45 2018

@author: USER
"""
# L 69 DECISION TREE 
# L70 

#copypasta from template

# Regression Template

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Position_Salaries.csv')
X = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2].values

# Splitting the dataset into the Training set and Test set - n/a b/c small set
"""from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)"""

# Feature Scaling
"""from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)
sc_y = StandardScaler()
y_train = sc_y.fit_transform(y_train)"""

# Fitting the DT Regression Model to the dataset !!! ---- !!!
from sklearn.tree import DecisionTreeRegressor
regressor = DecisionTreeRegressor(random_state = 0)
regressor.fit(X, y)
#gives output

# Predicting a new result
y_pred = regressor.predict(6.5)

# Visualising the Regression results -- using low res = problem
# red flag -- model seems to perfect.  so no actual predictions ...
# is predicting each of 10 levels, then joining by straight line 
# but this is noncontinuous model - change to higher res 

X_grid = np.arange(min(X), max(X), 0.01)
X_grid = X_grid.reshape((len(X_grid), 1))
plt.scatter(X, y, color = 'red')
plt.plot(X_grid, regressor.predict(X_grid), color = 'blue')
plt.title('Truth or Bluff (DTR Model)')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()

File: 2 Regression/E Decision Tree/my_decision_tree_regression.R
# lecture 72 - DT in R

#copypasta
# Regression Template

# Importing the dataset
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3]

# Splitting the dataset into the Training set and Test set
# # install.packages('caTools')
# library(caTools)
# set.seed(123)
# split = sample.split(dataset$Salary, SplitRatio = 2/3)
# training_set = subset(dataset, split == TRUE)
# test_set = subset(dataset, split == FALSE)

# Feature Scaling - should not be needed in DT b/c DT n/ based on Euclildian distance
# training_set = scale(training_set)
# test_set = scale(test_set)

# Fitting the DT Regression Model to the dataset
install.packages('rpart')
library(rpart)

# needed to add control parameter to get more than one split
# minsplit = 1 gives about four splits, another trap ... -- see notes under 1st ggplot
# trap is resolution is n/ high enough, connecting lines are artefacts 
# (DT is not continuous -- need to change visualization)
regressor = rpart(formula = Salary ~ .,
                  data = dataset,
                  control = rpart.control(minsplit = 1)
                  )   

# Predicting a new result
y_pred = predict(regressor, data.frame(Level = 6.5))

# Visualising the Regression Model results
# install.packages('ggplot2')
library(ggplot2)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') +
  geom_line(aes(x = dataset$Level, y = predict(regressor, newdata = dataset)),
            colour = 'blue') +
  ggtitle('Truth or Bluff (DT Regression Model)') +
  xlab('Level') +
  ylab('Salary')

# straight horizontal line comes out similar to SVR in Py (there neeeded FS)
# problem = not enough conditions / splits (no splits = all 1 prediction)

# Visualising the Regression Model results (for higher resolution and smoother curve)
# install.packages('ggplot2')
library(ggplot2)
x_grid = seq(min(dataset$Level), max(dataset$Level), 0.1)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') +
  geom_line(aes(x = x_grid, y = predict(regressor, newdata = data.frame(Level = x_grid))),
            colour = 'blue') +
  ggtitle('Truth or Bluff (DT Regression Model)') +
  xlab('Level') +
  ylab('Salary')